relay-event-schema/src/protocol/contexts/trace.rs

```diff
@@ -113,45 +114,76 @@ impl IntoValue for TraceId {
         Self: Sized,
         S: Serializer,
     {
-        Serialize::serialize(&self.to_string(), s)
+        s.collect_str(self)
     }
 }
 
-/// A 16-character hex string as described in the W3C trace context spec.
-#[derive(
-    Clone, Debug, Default, Eq, Hash, PartialEq, Ord, PartialOrd, Empty, IntoValue, ProcessValue,
-)]
-pub struct SpanId(pub String);
+/// A 16-character hex string as described in the W3C trace context spec, stored
+/// internally as an array of 8 bytes.
+#[derive(Clone, Copy, Default, Eq, Hash, PartialEq, Ord, PartialOrd)]
+pub struct SpanId([u8; 8]);
 
 relay_common::impl_str_serde!(SpanId, "a span identifier");
 
 impl FromStr for SpanId {
     type Err = Error;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(SpanId(s.to_string()))
+        match u64::from_str_radix(s, 16) {
+            Ok(id) if s.len() == 16 && id > 0 => Ok(Self(id.to_be_bytes())),
+            _ => Err(Error::invalid("not a valid span id")),
+        }
+    }
+}
+
+impl TryFrom<&[u8]> for SpanId {
+    type Error = Error;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        match <[u8; 8]>::try_from(value) {
+            Ok(bytes) if !bytes.iter().all(|&x| x == 0) => Ok(Self(bytes)),
+            _ => Err(Error::invalid("not a valid span id")),
+        }
+    }
+}
+
+impl fmt::Debug for SpanId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        struct DebugBytes([u8; 8]);
+
+        impl fmt::Debug for DebugBytes {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                write!(f, "\"")?;
+                for b in self.0 {
+                    write!(f, "{b:02x}")?;
+                }
+                write!(f, "\"")
+            }
+        }
+        f.debug_tuple("SpanId").field(&DebugBytes(self.0)).finish()
```

Comment:
Maybe we want the same here like we have for `TraceId`, just a simple inline format, so debug snapshots are a bit more concise?